ipaserver/plugins/cert.py

```diff
@@ -1046,6 +1054,84 @@ def _principal_name_matches_principal(name, principal_obj):
     return principal in principal_obj.get('krbprincipalname', [])
 
 
+def _validate_san_ips(generalnames):
+    """
+    Raise a ValidationError if the subjectAltName in a CSR includes
+    any IP address(es) that do not match a DNS name in the SAN.
+
+    """
+    san_dns_ips = set()
+    for name in generalnames:
+        if isinstance(name, cryptography.x509.general_name.DNSName):
+            san_dns_ips.update(_san_dnsname_ips(name.value))
+    for name in generalnames:
+        if isinstance(name, cryptography.x509.general_name.IPAddress):
+            if unicode(name.value) not in san_dns_ips:
+                raise errors.ValidationError(
+                    name='csr',
+                    error=_(
+                        "IP address in subjectAltName (%s) does not "
+                        "match any DNS name"
+                    ) % name.value
+                )
+
+
+def _san_dnsname_ips(dnsname, dnsname_is_cname=False):
+    """
+    Returns a set of IP addresses, managed by this IPa instance,
+    that correspond to the DNS name (from the subjectAltName).
+
+    """
+    fqdn = dnsutil.DNSName(dnsname).make_absolute()
+    if fqdn.__len__() < 4:
```

Comment:
this can be
```python
if len(fqdn) < 4:
```

Why is this required?  It is not clear to me why a name consisting of fewer than 4 components
cannot be resolved.

Reply:
The goal is to avoid attempting to "resolve" relative DNS names and DNS names for top-level domains.  Keep in mind that the "root" element of a DNS name counts, so a name like www.foo.com has a length of 4.

Reply:
I don't think we can make this assumption.  With gTLDs a customer really could have a domain name of "nike." or "coke." or whatever.  And maybe they want an IPAddress certificate for it... seems unlikely but I just don't see any reason to impose this restriction, given that we're only use A/AAAA records managed by this IPA deployment for verification.

Reply:
OK.  So I can reduce the threshold to 3, which will allow names like "shoes.nike." (len = 3), while still ruling out completely unqualified names/TLDs.  Or I can completely remove the check.  IIRC, this will mean waiting for DNS timeouts when attempting to resolve the zone for unqualified names.

Any preference?

Reply:
We do not allow single label deployments of IPA, at least not with trust to AD. I think having at least two labels limitation here is sensible.

Reply:
Someone will eventually want to deploy IPA for the gTLD.  Let's avoid the restriction here, even if it is currently imposed elsewhere.  After all, we will still check that IPA DNS has authority over the name and IP address (PTR record) in question.

Reply:
Hi @ipilcher , I'm back after a week of vacation and will be turning my attention back to this PR very soon (probably tomorrow).

Reply:
@ipilcher how does IPA currently issue a certificate with an unqualified name?  Does it not require creating a host or service principal with that exact (unqualified) name?  I cannot see any other way.

Is it a common practice to have hosts/services with unqualified names in the IPA deployment?  I suspect not.

Reply:
As far as I can tell, it "just works."  For example, I've created a host for one of my switches - ``switch4.penurio.us``, and I'm able to issue a certificate with both ``switch4.penurio.us`` and ``switch4`` as DNS names in the subjectAltName.

I used this CSR:

```
-----BEGIN CERTIFICATE REQUEST-----
MIIBqzCCARQCAQAwMjETMBEGA1UECgwKUEVOVVJJTy5VUzEbMBkGA1UEAwwSc3dp
dGNoNC5wZW51cmlvLnVzMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDE60MB
/fkVDwvAF3aSw5EkfucNloipohXBVD9T6dY01yEIVvUvUX1eraloGUPyZa+SUMyH
/qT6fOwzUen00hHl3fUHcWr68fCQHvzPXoYPSfMm2qHNRsAK0jRvyL75xhiBLTON
GtHvZ87xOoWKp3R/o0lsXyUU7uIubbJgzBeLoQIDAQABoDkwNwYJKoZIhvcNAQkO
MSowKDAmBgNVHREEHzAdghJzd2l0Y2g0LnBlbnVyaW8udXOCB3N3aXRjaDQwDQYJ
KoZIhvcNAQELBQADgYEAfyk6s0SZJ7vZsDnEf1wJG001FQnWnEvQz3KbLCaOaLbQ
RybUhz4EVFut+CgwgpKpZ+R8MST+ZLDBuToOkhdNEXWWktiyYriXRF8VXIVpRxKt
K5vHxMbPAcQwL9PK9EJSKFYKJUnAVSTpBCdOAu82WQB1SSi0O4sUNXs/GBn+OrA=
-----END CERTIFICATE REQUEST-----
```

And IPA (``python2-ipaserver-4.5.0-22.el7.centos.noarch``) issued this certificate:

```
-----BEGIN CERTIFICATE-----
MIIDszCCApugAwIBAgIEH/8AETANBgkqhkiG9w0BAQsFADA1MRMwEQYDVQQKEwpQ
RU5VUklPLlVTMR4wHAYDVQQDExVDZXJ0aWZpY2F0ZSBBdXRob3JpdHkwHhcNMTgw
NDExMTQwMTI5WhcNMjMwNDEyMTQwMTI5WjAyMRMwEQYDVQQKDApQRU5VUklPLlVT
MRswGQYDVQQDDBJzd2l0Y2g0LnBlbnVyaW8udXMwgZ8wDQYJKoZIhvcNAQEBBQAD
gY0AMIGJAoGBAMTrQwH9+RUPC8AXdpLDkSR+5w2WiKmiFcFUP1Pp1jTXIQhW9S9R
fV6tqWgZQ/Jlr5JQzIf+pPp87DNR6fTSEeXd9Qdxavrx8JAe/M9ehg9J8ybaoc1G
wArSNG/IvvnGGIEtM40a0e9nzvE6hYqndH+jSWxfJRTu4i5tsmDMF4uhAgMBAAGj
ggFQMIIBTDAfBgNVHSMEGDAWgBTeReJBnbTv7K95sQAnIHhqo24IXDA8BggrBgEF
BQcBAQQwMC4wLAYIKwYBBQUHMAGGIGh0dHA6Ly9pcGEtY2EucGVudXJpby51cy9j
YS9vY3NwMA4GA1UdDwEB/wQEAwIE8DAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYB
BQUHAwIwdQYDVR0fBG4wbDBqoDKgMIYuaHR0cDovL2lwYS1jYS5wZW51cmlvLnVz
L2lwYS9jcmwvTWFzdGVyQ1JMLmJpbqI0pDIwMDEOMAwGA1UECgwFaXBhY2ExHjAc
BgNVBAMMFUNlcnRpZmljYXRlIEF1dGhvcml0eTAdBgNVHQ4EFgQUikM/FyUSAV90
BPMjevjTu7Vt8X8wJgYDVR0RBB8wHYISc3dpdGNoNC5wZW51cmlvLnVzggdzd2l0
Y2g0MA0GCSqGSIb3DQEBCwUAA4IBAQAZjAREXMp9rTxZYYSNvI8m05M+2Qq5tiSp
CKUOaoDZAe/ivQOJLBucCwipGVT4TlnPsV3jvbcHbENw/U+BQEK21OJvIkOPXKjq
p3PpDVHw3fY/f8VP2uZ5OBueYbZNQYFFN1Ad7G17CcvoWTwLbtxwqQ6O8h0G4bKf
M8+bbcMxpQQh6UwpY2LG+07fBjC/zOX0BmHsK8fgF7ggC44TvZnG7x93Ie8MDBPy
LUQOcPqhkL5dnYkLk8TTm1Lkx8O13zlPw2iqYj3IRWjueMFWi1a68rHtl3os4ah3
EdtwHvOsWxVQNXeCwTxbC+wyoMilMg4cdNI2j+jtfIjLECyzFWVs
-----END CERTIFICATE-----
```

Reply:
@ipilcher interesting, thanks.  I did not realise this - it is not an intentional part of `cert-request` behaviour, but a side-effect of `host-find`, which appends the domain to unqualified names.

Thinking about a principled design that does not impose the name "depth limit" (even if other parts of IPA do)... maybe we can do it this way:

1. (fact of current behaviour) for each SAN DNSName we locate a matching host or service principal.  An unqualified DNSName will either resolve to a principal entry whose qualified name we then know (because it appears in the entry), or the request will be rejected (because no matching host/service entry was found).

2. Therefore, when producing the set of IP addresses corresponding to the DNSNames (`_san_dnsname_ips`), instead of iterating the SAN DNSNames, iterate the qualified hostnames of the principal entries produced at (1).

This will ensure we only try to resolve IP addresses of qualified names, while allowing the possibility of a single-element qualified name (e.g. `nike.`).  (Again, never mind that the rest of IPA still has  to catch up with this reality.  But I do not want to perpetuate this bad assumption in new code).

Reply:
That certainly makes sense (to the extent that I understand it).  I must admit that I have no clue how to actually implement it, though.

Reply:
@ipilcher I will be happy to implement this behaviour.  If you would push a new patch that fixes the tests and implements the other nits/suggestions already discussed, I will then implement and commentate the suggested behaviour and we can discuss and evaluate that implementation.

I do not mean to condescend or suggest that the approach cannot be explained so that you can implement it; rather let's collaborate on the code to un-stall your PR.

Reply:
First, I want to apologize for being so unresponsive on this.  Conference prep, day job, kids ...  :-(
> I do not mean to condescend or suggest that the approach cannot be explained so that you can implement it; rather let's collaborate on the code to un-stall your PR.
Absolutely no offense taken.  I'd just like to get this feature working.
I believe that I've just pushed an updated PR that addresses all of the points that have been raised.
Thanks for your patience!